File: src/interfaces_pkg/msg_codec_cpp/src/msg_publisher.cpp
#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/u_int8_multi_array.hpp"
#include "../include/msg_codec_cpp/msg_coder.hpp"
#include <cctype>
#include <iomanip>

using MType = MsgCoder::ValType;

class MsgPublisher : public rclcpp::Node {
public:
    MsgPublisher() : Node("msg_publisher") {
        MsgCoder::registerCommand(std::string("MV"), {{"cmd", MType::CHAR2}, {"id", MType::INT8}, {"pos", MType::INT64}, {"vel", MType::DOUBLE}});
        MsgCoder::registerCommands({
            {std::array<char, 2>{'C', 'F'}, {{"cmd", MType::CHAR2}, {"id", MType::INT8}, {"pos", MType::INT64}, {"vel", MType::DOUBLE}}},
            {std::array<char, 2>{'R', 'S'}, {{"cmd", MType::CHAR2}, {"id", MType::INT8}, {"status", MType::INT8}}}
        });

        publisher_ = this->create_publisher<std_msgs::msg::UInt8MultiArray>("encoded_msg", 10);
        timer_ = this->create_wall_timer(
            std::chrono::seconds(1),
            std::bind(&MsgPublisher::publish_message, this)
        );
        RCLCPP_INFO(this->get_logger(), "MsgPublisher has been started.");
    }

private:
    MsgCoder::KeyValues data_to_encode = {
        {"cmd", std::array<char, 2>{'M', 'V'}},
        {"id",  static_cast<int8_t>(6)},
        {"pos", static_cast<int64_t>(1000)},
        {"vel", static_cast<double>(3.14159)}
    };
    void publish_message() {
        int64_t val = std::get<int64_t>(data_to_encode["pos"]);
        data_to_encode["pos"] = val + 1;

        try {
            auto encoded_msg = MsgCoder::msg_encode(data_to_encode);
            std_msgs::msg::UInt8MultiArray msg;
            msg.data = std::vector<uint8_t>(encoded_msg.begin(), encoded_msg.end());
            publisher_->publish(msg);

            std::ostringstream oss;
            for (const auto& byte : encoded_msg) {
                if (std::isprint(byte)) {
                    oss << byte;
                } else {
                    oss << "\\x" << std::hex << std::setw(2) << std::setfill('0') << (static_cast<int>(byte) & 0xFF);
                }
            }
            RCLCPP_INFO(this->get_logger(), "Published encoded message: %s", oss.str().c_str());
        } catch (const std::exception &e) {
            RCLCPP_ERROR(this->get_logger(), "Error encoding message: %s", e.what());
        }
    }

    rclcpp::Publisher<std_msgs::msg::UInt8MultiArray>::SharedPtr publisher_;
    rclcpp::TimerBase::SharedPtr timer_;
};

int main(int argc, char *argv[]) {
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<MsgPublisher>());
    rclcpp::shutdown();
    return 0;
}

File: src/interface_pkgs/interfaces_pkg/msg_codec_py/msg_codec_py/msg_publisher.py
import rclpy
from rclpy.node import Node
from std_msgs.msg import UInt8MultiArray
from msg_codec_py.msg_coder import MsgCoder

class MsgPublisher(Node):
    def __init__(self):
        super().__init__('msg_publisher')
        MsgCoder.add_cmd_script(b'MV', {'cmd': '2s', 'id': 'b', 'pos': 'q', 'vel': 'd'})
        self.publisher_ = self.create_publisher(UInt8MultiArray, 'encoded_msg', 10)
        self.timer = self.create_timer(1.0, self.publish_message)
        self.get_logger().info('MsgPublisher has been started.')
        self.data_to_encode = {
            'cmd': b'MV',
            'id': 6,
            'pos': 1000,
            'vel': 3.1415926
        }

    def publish_message(self):
        self.data_to_encode['pos'] += 1  # 模拟位置变化
        try:
            encoded_msg = MsgCoder.msg_encode(self.data_to_encode)
            msg = UInt8MultiArray()
            msg.data = list(encoded_msg)
            self.publisher_.publish(msg)
            self.get_logger().info(f'Published encoded message: {encoded_msg}')
        except Exception as e:
            self.get_logger().error(f'Error encoding message: {e}')

def main(args=None):
    rclpy.init(args=args)
    node = MsgPublisher()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        node.get_logger().info('Publisher stopped by user.')
    finally:
        node.destroy_node()
        rclpy.shutdown()

if __name__ == '__main__':
    main()